### PR TITLE
driver: gic: support sgi raise for gicv3

### PR DESCRIPTION
**Pull request by PeikanTsai**
Use icc_sgi1r and icc_asgi1r to send SGI for gicv3.

### CODE REVIEW

File: core/drivers/gic.c [hunk 1]
```diff
@@ -389,22 +389,40 @@ static void gic_it_set_pending(struct gic_data *gd, size_t it)
...
+			SHIFT_U64(mask_aff1, 16) |
+			SHIFT_U64(mask_id, 24) |
+			SHIFT_U64(mask_aff2, 32));
+#else
+	uint32_t mask_group = group & 0x1;
 	uint32_t mask = (mask_id | SHIFT_U32(mask_group, 15) |
 		SHIFT_U32(mask_cpu, 16));
+#endif
 
 	assert(gd == &gic_data);
```
Comment (jenswi-linaro) [inline core/drivers/gic.c:412]: This function is not easy to read with the last ifdefs. How about moving these asserts to the calling function and make this function:
``` C
{
#if defined(CFG_ARM_GICV3)
...
#else
...
#endif
}
```
with no code outside the ifdef?

File: core/drivers/gic.c [hunk 2]
```diff
@@ -389,22 +389,40 @@ static void gic_it_set_pending(struct gic_data *gd, size_t it)
...
+			     uint8_t cpu_mask, uint8_t group)
 {
 	uint32_t mask_id = it & 0xf;
-	uint32_t mask_group = group & 0x1;
 	uint32_t mask_cpu = cpu_mask & 0xff;
+#if defined(CFG_ARM_GICV3)
+	/* Only support sending SGI to the cores in the same cluster now */
+	uint64_t mpidr = read_mpidr();
+	uint64_t mask_aff1 = (mpidr & MPIDR_AFF1_MASK) >> MPIDR_AFF1_SHIFT;
+	uint64_t mask_aff2 = (mpidr & MPIDR_AFF2_MASK) >> MPIDR_AFF2_SHIFT;
```
Comment (jenswi-linaro) [inline core/drivers/gic.c:401]: Aff3 missing?

File: core/arch/arm/include/arm64.h [hunk 3]
```diff
@@ -484,6 +484,8 @@ DEFINE_REG_WRITE_FUNC_(icc_eoir0, uint32_t, S3_0_c12_c8_1)
 DEFINE_REG_WRITE_FUNC_(icc_eoir1, uint32_t, S3_0_c12_c12_1)
 DEFINE_REG_WRITE_FUNC_(icc_igrpen0, uint32_t, S3_0_C12_C12_6)
 DEFINE_REG_WRITE_FUNC_(icc_igrpen1, uint32_t, S3_0_C12_C12_7)
+DEFINE_REG_WRITE_FUNC_(icc_sgi1r, uint64_t, S3_0_C12_C11_5)
```
Comment (jenswi-linaro) [inline core/arch/arm/include/arm64.h:487]: Please make this change in a separate commit.

Comment (PeikanTsai) [reply core/arch/arm/include/arm64.h:487]: Got it. I'll push a separate commit for it.

File: core/drivers/gic.c (hunk 1 above)
Comment (PeikanTsai) [reply core/drivers/gic.c:412]: Agree. I'll update the patch as you suggested.

File: core/drivers/gic.c (hunk 2 above)
Comment (PeikanTsai) [reply core/drivers/gic.c:401]: I'll add it in the updated patch also, thanks!

**Review by jenswi-linaro — COMMENTED** [review]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by PeikanTsai** [comment]
@jenswi-linaro  Could you please submit theses patches? Thanks!

**Comment by jforissier** [comment]
@PeikanTsai merged, thanks.
